lib/cmd/lint.js

```diff
@@ -1,165 +1,276 @@
+
+  if (utils.isComponent(url)) {
+    return checkComponent(url, prefixes.getFromUrl(url), concurrency);
+  } else if (utils.isPage(url)) {
+    return checkPage(url, prefixes.getFromUrl(url), concurrency);
   } else {
-    logger.error('Please provide a url or file(s) to lint!');
+    return checkPublicUrl(url, concurrency);
   }
 }
 
-module.exports = {
-  command: 'lint [url]',
-  describe: 'lint component data and schema',
-  aliases: ['l', 'linter'],
-  builder,
-  handler
-};
+/**
+ * determine if a schema has a description
+ * @param  {object} obj
+ * @return {boolean}
+ */
+function noDescription(obj) {
+  return !_.has(obj, '_description');
+}
+
+/**
+ * check camelCase, but more flexible than lodash's algorithm
+ * @param  {string}  str
+ * @return {Boolean}
+ */
+function isCamelCased(str) {
+  const camel = _.camelCase(str);
+
+  // note: we're lowercasing everything because we don't care about 100% perfect camelCase (e.g. imageURL is valid),
+  // we just want to make sure that people aren't using weird characters that cannot be used as properties in dot notation
+  return str.toLowerCase() === camel.toLowerCase() || str.toLowerCase() === `_${camel}`.toLowerCase(); // lodash's camelCase() will remove leading underscores, which we allow in things like _version, _description, and _devDescription
+}
+
+/**
+ * determine if a schema has camelCased props
+ * @param  {obj} obj
+ * @return {array}
+ */
+function nonCamelCasedProps(obj) {
+  return _.reduce(obj, (errors, value, key) => {
+    return !isCamelCased(key) ? errors.concat(key) : errors;
+  }, []);
+}
+
+/**
+ * determine if a schema has groups that reference non-existant fields
+ * @param  {obj} obj
+ * @return {array}
+ */
+function nonexistentGroupFields(obj) {
+  return _.reduce(_.get(obj, '_groups'), (errors, group, groupName) => {
+    const fields = group.fields;
+
+    _.each(fields, (field) => {
+      if (!_.has(obj, field)) {
+        errors.push(`${groupName} » ${field}`);
+      }
+    });
+    return errors;
+  }, []);
+}
+
+/**
+ * lint schemas for:
+ * - valid yaml syntax
+ * - has _description
+ * - all root-level properties are camelCase
+ * - _group fields refer to existing properties
+ * @param  {string} str of yaml
+ * @return {Stream}
+ */
+function lintSchema(str) {
+  return h([str])
+    .map(yaml.safeLoad)
+    .errors((e, push) => {
+      push(null, { type: 'error', message: `YAML syntax error: ${e.message.slice(0, e.message.indexOf(':'))}` });
+    }).flatMap((obj) => {
+      if (obj.type && obj.type === 'error') {
+        return h.of(obj); // pass through errors
+      } else {
+        let errors = [];
+
+        // lint schema!
+        if (noDescription(obj)) {
+          errors.push({ type: 'error', message: 'Schema has no _description' });
+        }
+
+        if (nonCamelCasedProps(obj).length) {
+          errors.push({ type: 'error', message: 'Properties must be camelCased', details: nonCamelCasedProps(obj).join('\n') });
```

Comment:
Just curious why this is enforced as (loose) camelCase vs just restricting to alpha characters?

Reply:
Properties can begin with underscores (e.g. `_description`)